admission-policy/readme.adoc

```diff
@@ -0,0 +1,342 @@
+openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days 100000 -extensions v3_req -extfile server.conf
+```
+
+=== 2.2: Deploy OPA on Kubernetes
+
+First, create a namespace to deploy OPA into.
+
+```bash
+kubectl create namespace opa
+```
+
+Create a Service to expose the OPA API. The Kubernetes API server will lookup
+the Service and execute webhook requests against it.
+
+**opa-admission-controller-service.yaml**:
+
+```yaml
+kind: Service
+apiVersion: v1
+metadata:
+  name: opa
+spec:
+  selector:
+    app: opa
+  ports:
+  - name: https
+    protocol: TCP
+    port: 443
+    targetPort: 443
+```
+
+```bash
+kubectl create -f opa-admission-controller-service.yaml -n opa
+```
+
+Next, create Secrets containing the TLS credentials for OPA:
+
+```bash
+kubectl create secret generic opa-ca --from-file=ca.crt -n opa
+kubectl create secret tls opa-server --cert=server.crt --key=server.key -n opa
+```
+
+Finally, create the Deployment to run OPA as an Admission Controller.
+
+**opa-admission-controller-deployment.yaml**:
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  labels:
+    app: opa
+  name: opa
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: opa
+      name: opa
+    spec:
+      containers:
+        - name: opa
+          image: openpolicyagent/opa:0.5.13
+          args:
+            - "run"
+            - "--server"
+            - "--tls-cert-file=/certs/tls.crt"
+            - "--tls-private-key-file=/certs/tls.key"
+            - "--addr=0.0.0.0:443"
+            - "--insecure-addr=127.0.0.1:8181"
+          volumeMounts:
+            - readOnly: true
+              mountPath: /certs
+              name: opa-server
+        - name: kube-mgmt
+          image: openpolicyagent/kube-mgmt:0.5
+          args:
+            - "--replicate=v1/pods"
+            - "--pod-name=$(MY_POD_NAME)"
+            - "--pod-namespace=$(MY_POD_NAMESPACE)"
+            - "--register-admission-controller"
+            - "--admission-controller-ca-cert-file=/certs/ca.crt"
+            - "--admission-controller-service-name=opa"
+            - "--admission-controller-service-namespace=$(MY_POD_NAMESPACE)"
+          volumeMounts:
+            - readOnly: true
+              mountPath: /certs
+              name: opa-ca
+          env:
+            - name: MY_POD_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            - name: MY_POD_NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+      volumes:
+        - name: opa-server
+          secret:
+            secretName: opa-server
+        - name: opa-ca
+          secret:
+            secretName: opa-ca
+```
+
+```bash
+kubectl create -f opa-admission-controller-deployment.yaml -n opa
+```
+
+When OPA starts, the sidecar (`kube-mgmt`) will register it as an External
```

Comment:
What is the purpose of this side car?

Reply:
kube-mgmt knows how to do k8s-specific things with OPA.  In this example, we use it just to register OPA as an external (OPA itself is general-purpose and can be used with any other system like k8s, istio, Linux, Terraform, etc.).  Here we use kube-mgmt to register OPA as an External Admission Controller with k8s.

Reply:
Adding a brief explanation in the doc would be helpful

Reply:
Yep--see the latest commit.